src/connectors/impls/gcl/writer/meta.rs

```diff
@@ -0,0 +1,604 @@
+                referer: http_request
+                    .get("referer")
+                    .as_str()
+                    .unwrap_or("")
+                    .to_string(),
+                latency: match http_request.get("latency").as_u64().unwrap_or(0) {
+                    0 => None,
+                    otherwise => Some(std::time::Duration::from_nanos(otherwise).into()),
+                },
+                cache_lookup: http_request.get("cache_lookup").as_bool().unwrap_or(false),
+                cache_hit: http_request.get("cache_hit").as_bool().unwrap_or(false),
+                cache_validated_with_origin_server: http_request
+                    .get("cache_validated_with_origin_server")
+                    .as_bool()
+                    .unwrap_or(false),
+                cache_fill_bytes: http_request.get("cache_fill_bytes").as_i64().unwrap_or(0),
+                protocol: http_request
+                    .get("protocol")
+                    .as_str()
+                    .unwrap_or("")
+                    .to_string(),
+            });
+        }
+    }
+
+    None
+}
+
+pub(crate) fn default_labels(config: &Config) -> StdHashMap<String, String> {
+    let mut labels = StdHashMap::new();
+
+    // Copy over default label entries
+    if let Some(config_labels) = &config.labels {
+        for (k, v) in config_labels {
+            labels.insert(k.clone(), v.to_string());
+        }
+    }
+
+    labels
+}
+
+pub(crate) fn labels(meta: Option<&HashMap<Cow<str>, Value>>) -> StdHashMap<String, String> {
+    let mut labels = StdHashMap::new();
+
+    // Override with metadata label entries
+    if let Some(has_meta) = meta {
+        if let Some(Value::Object(meta_labels)) = has_meta.get("labels") {
+            for (k, v) in meta_labels.iter() {
+                labels.insert(k.to_string(), v.to_string());
+            }
+        }
+    }
+
+    labels
+}
+
+pub(crate) fn operation(meta: Option<&HashMap<Cow<str>, Value>>) -> Option<LogEntryOperation> {
+    // Override for a specific per event trace
+    if let Some(has_meta) = meta {
+        if let Some(operation @ Value::Object(_)) = has_meta.get("operation") {
+            return Some(LogEntryOperation {
+                id: operation.get("id").as_str().unwrap_or("").to_string(),
+                producer: operation.get("producer").as_str().unwrap_or("").to_string(),
+                first: operation.get_bool("first").unwrap_or(false),
+                last: operation.get_bool("last").unwrap_or(false),
+            });
+        }
+    }
+
+    // Otherwise, None as mapping is optional
+    None
+}
+
+pub(crate) fn trace(meta: Option<&HashMap<Cow<str>, Value>>) -> String {
+    // Override for a specific per event trace
+    if let Some(has_meta) = meta {
+        if let Some(trace) = has_meta.get("trace") {
+            return trace.to_string();
+        }
+    }
+
+    // NOTE user supplied defaults supported at this time
+
+    "".to_string()
+}
```

Comment:
Passing in `Option<&Value>` could simplify this and the other functions a lot:
```suggestion
pub(crate) fn get_key_or_dflt(meta: Option<&Value>, key: &str) -> String {

    // Override for a specific per event trace
    // NOTE user supplied defaults supported at this time
    meta.get_str(key).unwrap_or_default().to_string()
}
```

Reply:
This was a good change, and it simplified some tests and removed some untested branches for the win